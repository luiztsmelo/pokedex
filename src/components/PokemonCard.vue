<template>
  <div ref="pokemonCard" class="pokemon-card" @click="pokemonCardHandleClick">
    <CardFront :pokemon="pokemon" v-if="routeQuery !== pokemon.name"/>
    <CardBack v-else/>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import CardFront from '@/components/CardFront.vue';
import CardBack from '@/components/CardBack.vue';

export default Vue.extend({
  props: {
    pokemon: { type: Object, required: true },
  },
  components: {
    CardFront,
    CardBack,
  },
  mounted(): void {
    // if this pokemon is in the query at component mount time, flip it
    if (this.routeQuery === this.pokemon.name) this.flipCard();
  },
  methods: {
    flipCard(): void {
      this.$refs.pokemonCard.style.transform = 'rotateY(180deg)';
    },
    unflipCard(): void {
      this.$refs.pokemonCard.style.transform = 'rotateY(0deg)';
    },
    pokemonCardHandleClick(): void {
      if (this.routeQuery === this.pokemon.name) {
        this.$router.push({ path: '/' });
        this.unflipCard();
      } else {
        this.$router.push({ path: '/', query: { q: this.pokemon.name } });
        this.flipCard();
      }
    },
  },
  computed: {
    routeQuery(): string | undefined {
      return this.$route.query.q;
    },
  },
  watch: {
    routeQuery(query) {
      if (query !== this.pokemon.name) this.unflipCard();
    },
  },
});
</script>

<style lang="scss">
.pokemon-card {
  cursor: pointer;
  overflow: hidden;
  height: 100%;
  border-radius: 20px;
  box-shadow: 3px 3px 8px 1px rgba(0,0,0,.16);
  transition: cubic-bezier(0.455, 0.03, 0.515, 0.955) all .22s;
}
.pokemon-card:hover {
  box-shadow: 3px 3px 20px 1px rgba(0,0,0,.35);
}
</style>
