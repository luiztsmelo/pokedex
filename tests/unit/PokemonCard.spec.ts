import { shallowMount, createLocalVue  } from '@vue/test-utils';
import VueRouter from 'vue-router';
import PokemonCard from '@/components/PokemonCard.vue';

const localVue = createLocalVue();
localVue.use(VueRouter);
const router = new VueRouter();

describe('PokemonCard.vue', () => {
  const wrapper = shallowMount(PokemonCard, {
    localVue,
    router,
    propsData: {
      pokemon: {
        id: 'UG9rZW1vbjowMDE=',
        number: '001',
        name: 'Bulbasaur',
        types: [
          'Grass', 'Poison',
        ],
      },
    },
  });
  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.vm.$route.query.q).toBe(undefined);
    expect(pokemonCardDiv.attributes().style).toBe(undefined);
  });
  const pokemonCardDiv = wrapper.find('.pokemon-card');
  test('click on unflipped pokemon card works fine', () => {
    pokemonCardDiv.trigger('click');
    expect(wrapper.vm.$route.query.q).toBe('Bulbasaur');
    expect(pokemonCardDiv.attributes().style).toBe('transform: rotateY(180deg);');
  });
  test('click on flipped pokemon card works fine', () => {
    pokemonCardDiv.trigger('click');
    expect(wrapper.vm.$route.query.q).toBe(undefined);
    expect(pokemonCardDiv.attributes().style).toBe(undefined);
  });
});
