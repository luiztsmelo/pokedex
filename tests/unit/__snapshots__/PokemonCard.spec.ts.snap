// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard.vue renders correctly 1`] = `
<div
  class="pokemon-card"
>
  <cardfront-stub
    pokemon="[object Object]"
  />
</div>
`;
